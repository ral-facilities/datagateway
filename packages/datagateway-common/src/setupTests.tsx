/* eslint-disable @typescript-eslint/no-empty-function */
import '@testing-library/jest-dom';
import React from 'react';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { initialState } from './state/reducers/dgcommon.reducer';
import { setLogger } from 'react-query';
import { WrapperComponent } from '@testing-library/react-hooks';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { createMemoryHistory, History } from 'history';
import failOnConsole from 'jest-fail-on-console';

failOnConsole();

jest.setTimeout(20000);

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {},
  });
}

// Add in ResizeObserver as it's not in Jest's environment
global.ResizeObserver = require('resize-observer-polyfill');

if (!global.structuredClone) {
  // structuredClone not available in jest/node <17, so this is a quick polyfill that should do the exact same thing
  global.structuredClone = (obj: unknown) => JSON.parse(JSON.stringify(obj));
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({ dgcommon: initialState });
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

// silence react-query errors
setLogger({
  log: console.log,
  warn: console.warn,
  error: jest.fn(),
});

// mock retry function to ensure it doesn't slow down query failure tests
jest.mock('./api/retryICATErrors', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(false),
}));

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

export const createReactQueryWrapper = (
  history: History = createMemoryHistory()
): WrapperComponent<unknown> => {
  const testQueryClient = createTestQueryClient();
  const state = {
    dgcommon: {
      ...initialState,
      facilityName: 'TEST',
      urls: {
        apiUrl: 'https://example.com/api',
        icatUrl: 'https://example.com/icat',
        idsUrl: 'https://example.com/ids',
        downloadApiUrl: 'https://example.com/topcat',
      },
    },
  };

  const mockStore = configureStore([thunk]);

  const wrapper: WrapperComponent<unknown> = ({ children }) => (
    <Provider store={mockStore(state)}>
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          {children}
        </QueryClientProvider>
      </Router>
    </Provider>
  );
  return wrapper;
};

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};
