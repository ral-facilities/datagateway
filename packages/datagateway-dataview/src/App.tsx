import { ConnectedRouter, routerMiddleware } from 'connected-react-router';
import {
  DGCommonMiddleware,
  DGThemeProvider,
  listenToMessages,
  MicroFrontendId,
  Preloader,
  BroadcastSignOutType,
  RequestPluginRerenderType,
} from 'datagateway-common';
import {
  createBrowserHistory,
  LocationListener,
  Location,
  Action,
} from 'history';
import * as log from 'loglevel';
import React from 'react';
import { Translation } from 'react-i18next';
import { batch, connect, Provider } from 'react-redux';
import { AnyAction, applyMiddleware, compose, createStore, Store } from 'redux';
import { createLogger } from 'redux-logger';
import thunk, { ThunkDispatch } from 'redux-thunk';
import './App.css';
import { saveApiUrlMiddleware } from './page/idCheckFunctions';
import PageContainer from './page/pageContainer.component';
import { configureApp } from './state/actions';
import { StateType } from './state/app.types';
import AppReducer from './state/reducers/app.reducer';
import { QueryClient, QueryClientProvider } from 'react-query';
import { ReactQueryDevtools } from 'react-query/devtools';

const history = createBrowserHistory();

// fix query string freeze bug
// see https://github.com/supasate/connected-react-router/issues/311#issuecomment-692017995
let listeners: LocationListener<unknown>[] = [];

function appendListener(fn: LocationListener<unknown>): () => void {
  let isActive = true;

  const listener: LocationListener<unknown> = (...args) => {
    if (isActive) fn(...args);
  };

  listeners.push(listener);

  return () => {
    isActive = false;
    listeners = listeners.filter((item) => item !== listener);
  };
}

function notifyListeners(
  ...args: [location: Location<unknown>, action: Action]
): void {
  listeners.forEach((listener) => listener(...args));
}

// make only one subscription to history changes and proxy to our internal listeners
history.listen((...args) => {
  // here's the key change
  batch(() => {
    notifyListeners(...args);
  });
});

// monkey patch to store subscriptions into our own pool
history.listen = (fn) => {
  return appendListener(fn);
};

const middleware = [
  thunk,
  routerMiddleware(history),
  DGCommonMiddleware,
  saveApiUrlMiddleware,
];

if (process.env.NODE_ENV === `development`) {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const logger = (createLogger as any)({ collapsed: true });
  middleware.push(logger);
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const whyDidYouRender = require('@welldone-software/why-did-you-render');
  whyDidYouRender(React);
}

/* eslint-disable no-underscore-dangle, @typescript-eslint/no-explicit-any */
const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

/* eslint-enable */

function mapPreloaderStateToProps(state: StateType): { loading: boolean } {
  return {
    loading: !state.dgdataview.settingsLoaded,
  };
}

export const ConnectedPreloader = connect(mapPreloaderStateToProps)(Preloader);

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: true,
      staleTime: 300000,
    },
  },
});

document.addEventListener(MicroFrontendId, (e) => {
  const action = (e as CustomEvent).detail;
  if (action.type === BroadcastSignOutType) {
    queryClient.clear();
  }
});

class App extends React.Component<unknown, { hasError: boolean }> {
  store: Store;

  public constructor(props: unknown) {
    super(props);
    this.state = { hasError: false };
    this.handler = this.handler.bind(this);

    // set up store in constructor to isolate from SciGateway redux store: https://redux.js.org/recipes/isolating-redux-sub-apps
    this.store = createStore(
      AppReducer(history),
      composeEnhancers(applyMiddleware(...middleware))
    );

    listenToMessages(this.store.dispatch);

    const dispatch = this.store.dispatch as ThunkDispatch<
      StateType,
      null,
      AnyAction
    >;
    dispatch(configureApp());
  }

  handler(e: Event): void {
    // attempt to re-render the plugin if we get told to
    const action = (e as CustomEvent).detail;
    if (action.type === RequestPluginRerenderType) {
      this.forceUpdate();
    }
  }

  public componentDidMount(): void {
    document.addEventListener(MicroFrontendId, this.handler);
  }

  public componentWillUnmount(): void {
    document.removeEventListener(MicroFrontendId, this.handler);
  }

  public componentDidCatch(error: Error | null): void {
    this.setState({ hasError: true });
    log.error(`datagateway_dataview failed with error: ${error}`);
  }

  public render(): React.ReactNode {
    if (this.state.hasError) {
      return (
        <div className="error">
          <React.Suspense
            fallback={<Preloader loading={true}>Finished loading</Preloader>}
          >
            <div
              style={{
                padding: 20,
                background: 'red',
                color: 'white',
                margin: 5,
              }}
            >
              <Translation>{(t) => t('app.error')}</Translation>
            </div>
          </React.Suspense>
        </div>
      );
    } else
      return (
        <div className="App">
          <Provider store={this.store}>
            <ConnectedRouter history={history}>
              <QueryClientProvider client={queryClient}>
                <DGThemeProvider>
                  <ConnectedPreloader>
                    <React.Suspense
                      fallback={
                        <Preloader loading={true}>Finished loading</Preloader>
                      }
                    >
                      <PageContainer />
                    </React.Suspense>
                  </ConnectedPreloader>
                </DGThemeProvider>
                <ReactQueryDevtools initialIsOpen={false} />
              </QueryClientProvider>
            </ConnectedRouter>
          </Provider>
        </div>
      );
  }
}

export default App;
