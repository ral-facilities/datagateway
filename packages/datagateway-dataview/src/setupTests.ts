/* eslint-disable @typescript-eslint/no-empty-function */
import '@testing-library/jest-dom';
// Blob implementation in jsdom is not complete (https://github.com/jsdom/jsdom/issues/2555)
// blob-polyfill fills in the gap
import 'blob-polyfill';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { initialState as dgDataViewInitialState } from './state/reducers/dgdataview.reducer';
import { dGCommonInitialState } from 'datagateway-common';

// Unofficial React 17 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

function noOp(): void {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({
  dgdataview: dgDataViewInitialState,
  dgcommon: dGCommonInitialState,
});
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: (args: unknown) => unknown) => fn);

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};
