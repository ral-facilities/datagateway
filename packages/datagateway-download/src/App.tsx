import {
  DGThemeProvider,
  MicroFrontendId,
  Preloader,
  RequestPluginRerenderType,
} from 'datagateway-common';
import React, { Component } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { ReactQueryDevtools } from 'react-query/devtools';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

import ConfigProvider from './ConfigProvider';
import AdminDownloadStatusTable from './downloadStatus/adminDownloadStatusTable.component';

import DownloadTabs from './downloadTab/downloadTab.component';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: true,
      staleTime: 300000,
    },
  },
});

/**
 * A fallback component when the app fails to render.
 */
function ErrorFallback(): JSX.Element {
  return (
    <div className="error">
      <div
        style={{
          padding: 20,
          background: 'red',
          color: 'white',
          margin: 5,
        }}
      >
        Something went wrong...
      </div>
    </div>
  );
}

class App extends Component<unknown, { hasError: boolean }> {
  public constructor(props: unknown) {
    super(props);
    this.state = { hasError: false };

    this.handler = this.handler.bind(this);
  }

  handler(e: Event): void {
    // attempt to re-render the plugin if we get told to
    const action = (e as CustomEvent).detail;
    if (action.type === RequestPluginRerenderType) {
      // This is a temporary fix for the current issue with the tab indicator
      // not updating after the size of the page has been altered.
      // This is issue is being tracked by material-ui (https://github.com/mui-org/material-ui/issues/9337).
      setTimeout(() => {
        window.dispatchEvent(new CustomEvent('resize'));
      }, 125);
      this.forceUpdate();
    }
  }

  public componentDidMount(): void {
    document.addEventListener(MicroFrontendId, this.handler);
  }

  public componentWillUnmount(): void {
    document.removeEventListener(MicroFrontendId, this.handler);
  }

  public render(): React.ReactNode {
    return (
      <div className="App">
        <DGThemeProvider>
          <ConfigProvider>
            <QueryClientProvider client={queryClient}>
              <React.Suspense
                fallback={
                  <Preloader loading={true}>Finished loading</Preloader>
                }
              >
                <Router>
                  <Switch>
                    <Route path="/admin/download">
                      <AdminDownloadStatusTable />
                    </Route>
                    <Route path="/download">
                      <DownloadTabs />
                    </Route>
                  </Switch>
                </Router>
              </React.Suspense>
              <ReactQueryDevtools initialIsOpen={false} />
            </QueryClientProvider>
          </ConfigProvider>
        </DGThemeProvider>
      </div>
    );
  }
}

export default App;
export { ErrorFallback };
