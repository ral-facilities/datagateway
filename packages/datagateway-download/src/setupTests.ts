/* eslint-disable @typescript-eslint/no-empty-function */
import '@testing-library/jest-dom';

function noOp(): void {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: (args: unknown) => unknown) => fn);

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};
