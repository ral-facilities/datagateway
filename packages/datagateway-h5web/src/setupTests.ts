import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

function noOp(): void {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

export const flushPromises = (): Promise<void> => new Promise(setImmediate);

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: (args: unknown) => unknown) => fn);
