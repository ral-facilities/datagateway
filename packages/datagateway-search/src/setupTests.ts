/* eslint-disable @typescript-eslint/no-empty-function */
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Action } from 'redux';
import { StateType } from './state/app.types';
import { dGCommonInitialState } from 'datagateway-common';
import { initialState as dgSearchInitialState } from './state/reducers/dgsearch.reducer';

// Unofficial React 17 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

function noOp(): void {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
}

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp });
}

// these are used for testing async actions
export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getState = (): Partial<StateType> => ({
  dgsearch: dgSearchInitialState,
  dgcommon: dGCommonInitialState,
});
export const dispatch = (action: Action): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getState);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// Mock lodash.debounce to return the function we want to call.
jest.mock('lodash.debounce', () => (fn: (args: unknown) => unknown) => fn);

// Add in ResizeObserver as it's not in Jest's environment
global.ResizeObserver = require('resize-observer-polyfill');

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};
