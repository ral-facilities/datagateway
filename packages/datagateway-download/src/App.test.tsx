import { render } from '@testing-library/react';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import App, { ErrorFallback } from './App';
import { flushPromises } from './setupTests';

jest.mock('loglevel');
jest.mock('./ConfigProvider');

describe('App', () => {
  it('renders without crashing', async () => {
    const div = document.createElement('div');

    ReactDOM.render(<App />, div);

    await act(async () => {
      await flushPromises();
    });

    ReactDOM.unmountComponentAtNode(div);
  });
});

describe('ErrorFallback', () => {
  it('should should render an error message for when app fails catastrophically', () => {
    const { asFragment } = render(<ErrorFallback />);
    expect(asFragment()).toMatchSnapshot();
  });
});
