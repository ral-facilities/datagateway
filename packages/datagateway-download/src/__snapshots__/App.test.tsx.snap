// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorFallback should should render an error message for when app fails catastrophically 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <div
      style="padding: 20px; background: red; color: white; margin: 5px;"
    >
      Something went wrong...
    </div>
  </div>
</DocumentFragment>
`;
